ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -653,3 +668,103 @@ VirtualizationRootHandle ActiveProvider_FindForPath(const char* _Nonnull path)
     
     return matchingHandle;
 }
+
+bool VirtualizationRoots_ProcessMayAccessOfflineRoots(pid_t pid)
+{
+    bool result = false;
+    
+    RWLock_AcquireShared(s_virtualizationRootsLock);
+    {
+        for (uint32_t i = 0; i < s_offlineIOPIDCount; ++i)
+        {
+            if (s_offlineIOPIDs[i] == pid)
+            {
+                result = true;
+                break;
+            }
+        }
+    }
+    RWLock_ReleaseShared(s_virtualizationRootsLock);
+
+    return result;
+}
+
+bool VirtualizationRoots_AddOfflineIOProcess(pid_t pid)
+{
+    bool success = false;
+    
+    // Note: lock may temporarily be dropped within
+    RWLock_AcquireExclusive(s_virtualizationRootsLock);
+    {
+        while (s_offlineIOPIDCount >= s_maxOfflineIOPIDs && s_maxOfflineIOPIDs < UINT32_MAX)
+        {
+            // Array must be resized
+            uint32_t resizeArrayLength = MIN(MAX(1u, s_maxOfflineIOPIDs * 2u), UINT32_MAX);
```

Comment:
Related: I agree with what @jeschu1 mentioned in an earlier comment that we're not gaining much from the user experience side with a resizable array (and the resizable array does make things more complicated).  

Users won't have more than a handful of `gvfs` commands running at once and so if we had a fixed length array with something like 64 entries I don't think we'd ever exceed it and it would be OK to fail the offline request if we did.

cc: @jrbriggs for his thoughts as well